docs/src/userguide/nogil.rst

```diff
@@ -0,0 +1,160 @@
+.. _cython_and_gil:
+
+Cython and the GIL
+==================
+
+Python has a global lock (:term:`the GIL <Global Interpreter Lock or GIL>`)
+to ensure that data related to the Python interpreter is not corrupted.
+It is *sometimes* to release this lock in Cython when you are not
+accessing Python data.
+
+There are two occasions when you may want to release the
+GIL:
+
+#. Using :ref:`Cython's parallelism mechanism <parallel>`. The contents of a ``prange`` loop for example are required to be ``nogil``.
+
+#. If you want other (external) Python threads to be able to run at the same time.
+
+    #. if you have a large computationally/IO-intensive block that doesn't need the GIL then it may be "polite" to release it, just to benefit users of your code who want to do multi-threading. However, this is mostly useful rather than necessary.
+
+    #. (very, very occasionally) it's sometimes useful to briefly release the GIL with a short with ``nogil: pass`` block. This is because Cython doesn't release it spontaneously (unlike Python) so if you're waiting on another Python thread to complete a task, this can avoid deadlocks. This sub-point probably doesn't apply to you unless you're compiling GUI code with Cython.
+
+If neither of these two points apply then you probably do not need to release the GIL. The sort of Cython 
+code that can run without the GIL (no calls to Python, purely C-level numeric operations)
+is often the sort of code that runs efficiently. This sometimes gives people the impression that the
+inverse is true and the trick is releasing the GIL, rather than the actual code they’re running.
+Don’t be misled by this -- your (single-threaded) code will run the same speed with or without the GIL.
+
+Marking functions as able to run without the GIL
+------------------------------------------------
+
+You can mark a whole function (either a Cython function or an :ref:`external function <nogil>`) as
+``nogil`` by appending this to the function signature or by using the ``@cython.nogil`` decorator:
+
+.. tabs::
+
+    .. group-tab:: Pure Python
+    
+        .. code-block:: python
+
+            @cython.nogil(True)
+            @cython.cfunc
+            @cython.noexcept
+            def some_func() -> None:
+            ...
+
+    .. group-tab:: Cython
+    
+        .. code-block:: cython
+    
+            cdef void some_func() noexcept nogil:
+                ....
+
+Be aware that this does not release the GIL when calling the function. It merely indicates that
+a function is suitable for use when the GIL is released. It is also fine to call these functions
+while holding the GIL.
+
+In this case we've marked the function as ``noexcept`` to indicate that it cannot raise a Python
+exception. Be aware that a function with an ``except *`` exception specification (typically functions
```

Comment:
I was wondering as I wrote this: is this a big enough pitfall that we should emit a warning?

Reply:
A warning would suggest that users are doing something wrong. They might be, but it's actually perfectly reasonable to use `except * nogil`, so a warning seems annoying and misleading.

Reply:
Indeed - this was partly what I was thinking of when I suggested (https://github.com/cython/cython/issues/5311)

We have a little-used "message" feature I think?